src/embedding.rs

```diff
@@ -338,27 +429,28 @@ pub fn calculate_embeddings<T1, T2>(
 
 /// Provides matrix multiplication based on sparse matrix data.
 #[derive(Debug)]
-struct MatrixMultiplicator<T: SparseMatrixReader + Sync + Send, M: MatrixWrapper> {
+struct MatrixMultiplicator<M: MatrixWrapper> {
     dimension: usize,
     number_of_entities: usize,
     fixed_random_value: i64,
-    sparse_matrix_reader: Arc<T>,
+    sparse_matrix_reader: Arc<SparseMatrix>,
     _marker: PhantomData<M>,
+    init_method: InitMethod,
 }
 
-impl<T, M> MatrixMultiplicator<T, M>
+impl<M> MatrixMultiplicator<M>
 where
-    T: SparseMatrixReader + Sync + Send,
     M: MatrixWrapper,
 {
-    fn new(config: Arc<Configuration>, sparse_matrix_reader: Arc<T>) -> Self {
+    fn new(config: Arc<Configuration>, sparse_matrix_reader: Arc<SparseMatrix>) -> Self {
         let rand_value = config.seed.map(hash).unwrap_or(0);
         Self {
             dimension: config.embeddings_dimension as usize,
             number_of_entities: sparse_matrix_reader.get_number_of_entities() as usize,
             fixed_random_value: rand_value,
             sparse_matrix_reader,
             _marker: PhantomData,
+            init_method: config.init_method,
```

Comment:
Nit, I suggest we have _marker as last argument. 'Important' params should be first, weird quirks args like _marker should be last